wen-restart/src/last_voted_fork_slots_aggregate.rs

```diff
@@ -90,80 +123,132 @@ impl LastVotedForkSlotsAggregate {
+        Some(record)
+    }
+
+    // Return true if the message has already been saved, so we can skip the rest of the processing.
+    fn update_and_check_if_message_already_saved(
+        &mut self,
+        new_slots: RestartLastVotedForkSlots,
+        new_slots_vec: Vec<Slot>,
+    ) -> bool {
+        let from = &new_slots.from;
         let new_slots_set: HashSet<Slot> = HashSet::from_iter(new_slots_vec);
         let old_slots_set = match self.last_voted_fork_slots.insert(*from, new_slots.clone()) {
             Some(old_slots) => {
                 if old_slots == new_slots {
-                    return None;
+                    return true;
                 } else {
-                    HashSet::from_iter(old_slots.to_slots(self.root_slot))
+                    HashSet::from_iter(old_slots.to_slots(self.root_bank.slot()))
                 }
             }
             None => HashSet::new(),
         };
         for slot in old_slots_set.difference(&new_slots_set) {
             let entry = self.slots_stake_map.get_mut(slot).unwrap();
-            *entry = entry.saturating_sub(sender_stake);
-            if *entry < threshold_stake {
-                self.slots_to_repair.remove(slot);
+            if let Some(sender_stake) = Self::validator_stake_at_slot(&self.root_bank, slot, from) {
+                *entry = entry.saturating_sub(sender_stake);
+                let repair_threshold_stake = (Self::total_stake_at_slot(&self.root_bank, slot)
+                    .unwrap() as f64
+                    * self.repair_threshold) as u64;
+                if *entry < repair_threshold_stake {
+                    self.slots_to_repair.remove(slot);
+                }
             }
         }
         for slot in new_slots_set.difference(&old_slots_set) {
             let entry = self.slots_stake_map.entry(*slot).or_insert(0);
-            *entry = entry.saturating_add(sender_stake);
-            if *entry >= threshold_stake {
-                self.slots_to_repair.insert(*slot);
+            if let Some(sender_stake) = Self::validator_stake_at_slot(&self.root_bank, slot, from) {
+                *entry = entry.saturating_add(sender_stake);
+                let repair_threshold_stake = (Self::total_stake_at_slot(&self.root_bank, slot)
+                    .unwrap() as f64
+                    * self.repair_threshold) as u64;
+                if *entry >= repair_threshold_stake {
+                    self.slots_to_repair.insert(*slot);
+                }
             }
         }
-        Some(record)
+        false
     }
 
-    pub(crate) fn active_percent(&self) -> f64 {
-        let total_stake = self.epoch_stakes.total_stake();
-        let total_active_stake = self.active_peers.iter().fold(0, |sum: u64, pubkey| {
-            sum.saturating_add(Self::validator_stake(&self.epoch_stakes, pubkey))
-        });
-        total_active_stake as f64 / total_stake as f64 * 100.0
+    fn update_epoch_info_vec(&mut self) {
```

Comment:
This seems a bit overly complicated, seeing as there's only two possible epochs, the root bank epoch and root bank epoch + 1. Can we just initialized this epoch list on construction of `LastVotedForkSlotsAggregate`

Reply:
Sounds good, changed.